examples/deep_learning/robust_training.py

```diff
@@ -86,62 +111,77 @@ def accuracy(params, images, labels):
 logistic_loss = jax.vmap(loss.multiclass_logistic_loss)
 
 
+@jax.jit
 def loss_fun(params, l2_regul, images, labels):
   """Compute the loss of the network."""
   logits = net.apply({"params": params}, images)
   sqnorm = tree_util.tree_l2_norm(params, squared=True)
   loss_value = jnp.mean(logistic_loss(labels, logits))
   return loss_value + 0.5 * l2_regul * sqnorm
 
-train_ds = load_dataset("train", is_training=True, batch_size=128)
-test_ds = load_dataset("test", is_training=False, batch_size=1000)
+
+@jax.jit
+def fsgm_attack(image, label, params, l2_regul, epsilon=0.1):
+  """Fast sign-gradient attack on the L-infinity ball with radius epsilon.
+
+  Parameters:
+    image: array-like, input data for the CNN
+    label: integer, class label corresponding to image
+    params: tree, parameters of the model to attack
+    l2_regul: float, L2 regularization scale in the loss function
+    epsilon: float, radius of the L-infinity ball.
+
+  Returns:
+    perturbed_image: Adversarial image on the boundary of the L-infinity ball of radius
+      epsilon and centered at image.
+  """
+  # comppute gradient of the loss wrt to the image
+  grad = jax.grad(loss_fun, argnums=2)(params, l2_regul, image, label)
+  adv_image = image + epsilon * jnp.sign(grad)
+  # clip the image to ensure pixels are between 0 and 1
+  return jnp.clip(adv_image, 0, 1)
+
+train_ds, test_ds = load_datasets()
 
 # Initialize solver and parameters.
 solver = OptaxSolver(opt=optax.adam(1e-3), fun=loss_fun)
 rng = jax.random.PRNGKey(0)
 params = CNN().init(rng, jnp.ones([1, 28, 28, 1]))["params"]
 l2_regul = 1e-4
 
+n_epochs = 10
+batch_size = 128
+
 state = solver.init_state(params)
+rng = jax.random.PRNGKey(0)
+
+for epoch in range(n_epochs):
+  # Training
+
+  pbar = tqdm(enumerate(shuffle(train_ds, rng, batch_size=128)))
+  acc = []
+  adv_acc = []
+  for it, (images, labels) in pbar:
```

Comment:
```suggestion
  for images, labels in pbar:
```
this goes together with the previous suggestion